src/main/java/io/confluent/connect/jdbc/source/TimestampIncrementingCriteria.java

```diff
@@ -218,15 +224,40 @@ public TimestampIncrementingOffset extractValues(
    *
    * @param schema the record's schema; never null
    * @param record the record's struct; never null
+   * @param timestampGranularity defines the configured granularity of the timestamp field
    * @return the timestamp for this row; may not be null
    */
   protected Timestamp extractOffsetTimestamp(
       Schema schema,
-      Struct record
+      Struct record,
+      String timestampGranularity
   ) {
     caseAdjustedTimestampColumns.computeIfAbsent(schema, this::findCaseSensitiveTimestampColumns);
     for (String timestampColumn : caseAdjustedTimestampColumns.get(schema)) {
-      Timestamp ts = (Timestamp) record.get(timestampColumn);
+      Timestamp ts;
+      switch (timestampGranularity) {
+        case TIMESTAMP_GRANULARITY_LONG_NANOS:
+          ts = DateTimeUtils.toTimestamp((long) record.get(timestampColumn));
+          break;
+        case TIMESTAMP_GRANULARITY_STRING_NANOS:
+          try {
+            ts = DateTimeUtils.toTimestamp((String) record.get(timestampColumn));
+          } catch (NumberFormatException  e) {
+            throw new ConnectException(
+                "Invalid value for timestamp column with nanos-string granularity: "
+                    + record.get(timestampColumn)
+                    + e.getMessage());
+          }
+          break;
+        case TIMESTAMP_GRANULARITY_STRING_ISO_DATETIME:
+          ts = DateTimeUtils.toTimestampFromIsoDateTime(
+              (String) record.get(timestampColumn)
+          );
+          break;
+        default:
+          ts = (Timestamp) record.get(timestampColumn);
```

Comment:
This switch block could also be handled with a `timestampGranularity.convert` if we had the granularity exposed as enum, and a `convert` method on the enum.